theme/helpers.js

```diff
@@ -9,13 +13,81 @@ exports.Color = class {
         Object.keys(shades).forEach((shade) => (this[shade] = shades[shade]));
     }
 
-    shade(shade) {
+    alter(shade = 500) {
 
-        return this[shade];
+        return Color(this[shade]);
     }
 
     toString() {
 
         return this['500'];
     }
 };
+
+exports.paletteFromTheme = (theme) => {
+
+    const colors = Object.keys(theme).reduce((acc, key) => {
+
+        const regex = /^color-(?<color>\w+)-(?<transparent>\w+-)?(?<shade>\d{3})/;
+
+        Assert(regex.test(key), `Expected ${key} to match regex ${regex}`);
+
+        const { color, transparent, shade } = key.match(/^color-(?<color>\w+)-(?<transparent>\w+-)?(?<shade>\d{3})/).groups;
+
+        if (!(color in acc)) {
+            acc[color] = { transparent: {} };
+        }
+
+        if (transparent) {
+            acc[color].transparent[shade] = theme[key];
+        }
+        else {
+            acc[color][shade] = theme[key];
+        }
+
+        return acc;
+    }, {});
+
+    return Object.keys(colors).reduce((acc, color) => {
+
+        acc[color] = new exports.Color(colors[color]);
+        return acc;
+    }, {});
+};
+
+exports.spacing = (...args) => {
+
+    if (args.length === 1) {
+        Assert(!isNaN(args[0]), `Expected '${args[0]}' to be a number`);
+        return parseInt(args[0], 10) * BASE_SCALING_FACTOR;
+    }
+
+    const argToValue = (a) => {
+
+        Assert(a === 'auto' || !isNaN(a), `Expected '${a}' to be a number or 'auto'`);
+        return a === 'auto' ? a : `${parseInt(a, 10) * BASE_SCALING_FACTOR}px`;
```

Comment:
Why do we treat `auto` as a special case, but not other keywords e.g. `initial`, `inherit`? And why do we set values produced in this route to pixels, but numeric values can be set to whatever unit at the component level?

If this is all stuff bic'd in some form or another from Material UI and we're taking what they give us, I'll hush up :D more curious than anything, I'm pretty short on CSS

Reply:
My goal here was to mimic MUI's `spacing` function which is admittedly a bit weird with the px stuff. But I'm actually not sure how it handles inherit, initial, etc. I'll have to look into that!

Reply:
Ay, gotcha. Yea, I assumed there was a lot more going on here than I understood 😂  No idea if considering `initial` and `inherit` is a sensible thing either, just realized padding and margin aren't even inherited